pgjdbc/src/test/java/org/postgresql/test/jdbc2/BinaryTest.java

```diff
@@ -0,0 +1,282 @@
+   *
+   * @throws SQLException if a database error occurs
+   */
+  @Test
+  public void testCustomBinaryTypes() throws SQLException {
+    con = TestUtil.openDB();
+
+    PgConnection pgconn = (PgConnection) con;
+    QueryExecutor queryExecutor = pgconn.getQueryExecutor();
+
+    // define an oid of a binary type for testing, POINT is used here as it already exists in the
+    // database and requires no complex own type definition
+    int customTypeOid = Oid.POINT;
+    // the only disadvantage is that it is added automatically in the PgConnection, so be sure to
+    // remove type OID from predefined lists
+    queryExecutor.getBinaryReceiveOids().remove(customTypeOid);
+    queryExecutor.getBinarySendOids().remove(customTypeOid);
+    assertFalse(queryExecutor.useBinaryForReceive(customTypeOid));
+    assertFalse(queryExecutor.useBinaryForSend(customTypeOid));
+
+    // now define a custom type
+    pgconn.addDataType("point", TestCustomType.class);
+    // check if it was automatically added
+    assertTrue(queryExecutor.useBinaryForReceive(customTypeOid));
+    assertTrue(queryExecutor.useBinaryForSend(customTypeOid));
+
+    // INSERT INTO test_binary(id,name,geom) values(1,'Test',Point(1,2))
+    try (Statement st = con.createStatement()) {
+      // insert some data for testing
+      st.execute("INSERT INTO test_binary(id,name,geom) values(1,'Test',Point(1,2))");
+      // try to read it
+      try (ResultSet rs = st.executeQuery("SELECT geom FROM test_binary WHERE id=1")) {
+        assertTrue(rs.next());
+        Object o = rs.getObject(1);
+        assertNotNull(o);
+        // make sure type is correct
+        assertEquals(TestCustomType.class, o.getClass());
+        TestCustomType co = (TestCustomType) o;
+        // first try should not read binary as no prepared statement was used and it was not called
+        // multiple times
+        assertFalse("was read binary though it should not be", co.wasReadBinary());
+      }
+    }
+
+    TestCustomType co;
+    // now enforce using binary mode
+    try (PreparedStatement pst = con.prepareStatement("SELECT geom FROM test_binary WHERE id=?")) {
+      pst.setInt(1, 1);
+      // execute it 10 times to trigger binary transfer
+      for (int i = 0; i < 10; i++) {
+        try (ResultSet rs = pst.executeQuery()) {
+          assertTrue(rs.next());
+        }
+      }
+      // do the real check
+      try (ResultSet rs = pst.executeQuery()) {
+        assertTrue(rs.next());
+        Object o = rs.getObject(1);
+        assertNotNull(o);
+        // make sure type is correct
+        assertEquals(TestCustomType.class, o.getClass());
+        co = (TestCustomType) o;
+        // now binary transfer should be working
+        // first try should not read binary as no prepared statement was used and it was not called
+        // multiple times
+        assertTrue("was not read binary though it should not be", co.wasReadBinary());
```

Comment:
Frankly speaking, whenever you have code comments before "assert" call, consider putting them into the assert message.
Then the comment will be printed to the CI log, and it would help the analysis.